File: src/core/validators/Enumeration/EnumerationItemsMustBeUnique.cs
using System.Linq;
using MetaEd.Grammar.Antlr;

namespace MetaEd.Core.Validator.Enumeration
{
    public class EnumerationItemsMustBeUnique : ValidationRuleBase<MetaEdGrammar.EnumerationContext>
    {
        private static string[] DuplicateShortDescriptions(MetaEdGrammar.EnumerationContext context)
        {
            var shortDescriptions = context.enumerationItem().Select(x => x.shortDescription().GetText());

            //group and filter duplicates
            return shortDescriptions.GroupBy(x => x).Where(group => group.Count() > 1).Select(group => group.Key).ToArray();
        }

        public override bool IsValid(MetaEdGrammar.EnumerationContext context)
        {
            return !DuplicateShortDescriptions(context).Any();
        }

        public override string GetFailureMessage(MetaEdGrammar.EnumerationContext context)
        {
            var identifier = context.enumerationName().GetText();
            var duplicateShortDescriptions = DuplicateShortDescriptions(context);

            return string.Format("Enumeration '{0}' declares duplicate item{2} '{1}'.", identifier, string.Join("', '", duplicateShortDescriptions), duplicateShortDescriptions.Count() > 1 ? "s" : string.Empty);
        }
    }
}

File: src/core/validators/Enumeration/EnumerationItemsMustBeUnique.ts
import { ValidationRuleBase } from "../ValidationRuleBase";
export class EnumerationItemsMustBeUnique extends ValidationRuleBase<MetaEdGrammar.EnumerationContext>
{
    private static duplicateShortDescriptions(context: MetaEdGrammar.EnumerationContext): string[] {
        let shortDescriptions = context.enumerationItem().map(x => x.shortDescription().getText());
        return shortDescriptions.GroupBy(x => x).filter(group => group.Count() > 1).map(group => group.Key).ToArray();
    }
    public isValid(context: MetaEdGrammar.EnumerationContext): boolean {
        return EnumerationItemsMustBeUnique.duplicateShortDescriptions(context).length==0;
    }
    public getFailureMessage(context: MetaEdGrammar.EnumerationContext): string {
        let identifier = context.enumerationName().getText();
        let duplicateShortDescriptions = EnumerationItemsMustBeUnique.duplicateShortDescriptions(context);
        return `Enumeration '${identifier}' declares duplicate item${duplicateShortDescriptions.length > 1 ? "s" : ""} '${duplicateShortDescriptions.join(', ')}'.`;
    }
}
